lib/opFns.js

```diff
@@ -156,7 +156,7 @@ module.exports = {
       return new BN(0)
     }
 
-    return word.shrn((31 - pos.toNumber()) * 8).andln(0xff)
+    return new BN(word.shrn((31 - pos.toNumber()) * 8).andln(0xff))
```

Comment:
How was this bug not caught?

`andln` returns a Javascript integer.

@pirapira @winsvega do we have proper state tests for `BYTE`?

Reply:
it passes the relevant vm tests, fwiw:
```
ok 171 valid gas usage [file: byte10]
...
ok 297 valid gas usage [file: byte1]
ok 298 valid return value
```

Reply:
This PR passes it, `master` or both pass it?

It may be some combination of instructions which triggered this. @yann what was the actual issue you ran into?

Reply:
in remix we have to format the stack for each step to be compliant to the geth trace. 
Before js vm would return a list of buffer for the stack , but it has been changed to BN.
And we end up by having a mix of BN and integger (due to BYTE)

Reply:
master (unpatched) passes the relevant VM tests. I haven't tried the PR.

Reply:
@axic `BYTE` returns 0 to 255. I think all possible return values fit as a Javascript integer.

Reply:
If the tests are passing, maybe you would not want to merge that.
If so please close the PR, we will handle this on remix side.

Reply:
Merge or not merge, it's fine for me either way.  If tests are missing as @axic says, I want to add tests.

Reply:
ethereumjs-vm keeps all stack elements as an instance of bn.js (BN) and as such returning a Javascript integer is wrong.

Running any arithmetic opcode after `BYTE` should fail, because they expect an instance of BN. Most of the memory and state related opcodes are safe because they don't expect the top element to be a BN. This makes me think we should have test cases of the format `x y DUP1 BYTE ADD` to force calculations with the result.

Reply:
I see.  I never thought about how return values from instructions are used afterward.

Reply:
I created an issue in tests repo https://github.com/ethereum/tests/issues/448

Reply:
It would be nice if somehow we could include the hash of memory and the stack in the tests. That should catch a lot of things (but not all) tracing tries to catch.

Reply:
@yann300 the patch is correct and should be merged. All the stack items should be an instance of BN. Please report if you find any other discrepancy.

Reply:
@axic I'm not sure about the checking the stack contents.  Implementations are allowed not to compute stack elements that are never used.